### PR TITLE
MM-6482: Across teams channel auto complete

### PR DESCRIPTION
**Pull request by AshishDhama**
#### Summary
Across team channel auto complete

#### Ticket Link
Fixes https://github.com/mattermost/mattermost/issues/20606

#### Release Note
```release-note
* Channel auto completion will now return results across all teams.
```

### CODE REVIEW

**Comment by unode** [comment]
@AshishDhama many thanks for this and the cross reference from the older issue.

If I understand the code correctly the syntax is `~channel-name(team-name)`.
Would you be so kind to add this information to the description of this PR?

This will be relevant to answer #24132 .

File: server/public/model/channel_mentions.go [hunk 1]
```diff
@@ -26,3 +28,22 @@ func ChannelMentions(message string) []string {
 
 	return names
 }
+
+func ChannelMentionsAcrossTeams(message string) map[string][]string {
```
Comment (agnivade) [inline server/public/model/channel_mentions.go:32]: Let's add a test for this.

Also, can we add some docs to explain the format of the `map[string][]string`?

File: server/public/model/channel_mentions.go [hunk 2]
```diff
@@ -26,3 +28,22 @@ func ChannelMentions(message string) []string {
...
+	names := map[string][]string{}
+
+	if strings.Contains(message, "~") {
+		alreadyMentioned := make(map[string]bool)
+		for _, match := range channelMentionAcrossTeamsRegexp.FindAllString(message, -1) {
+			channelString := channelMentionRegexp.FindString(match)
+			teamString := channelMentionTeamRegexp.FindString(match)
```
Comment (agnivade) [inline server/public/model/channel_mentions.go:39]: Any reason we need to use regexes again to extract the channel and team names? We can just declare groups in the regex and use `FindAllStringSubmatch` to get the submatches as well in a single shot.

See: https://pkg.go.dev/regexp#pkg-overview for more info

> If 'Submatch' is present, the return value is a slice identifying the successive submatches of the expression. Submatches are matches of parenthesized subexpressions (also known as capturing groups) within the regular expression, numbered from left to right in order of opening parenthesis. Submatch 0 is the match of the entire expression, submatch 1 is the match of the first parenthesized subexpression, and so on.

For an example implementation, see: https://github.com/mattermost/mattermost-plugin-github/blob/11815f80ac05b110a5de479bbb7a9b1ff2995463/server/plugin/permalinks.go#L46

File: server/channels/app/post.go [hunk 3]
```diff
@@ -461,6 +461,7 @@ func (a *App) attachFilesToPost(post *model.Post) *model.AppError {
 // If channel is nil, FillInPostProps will look up the channel corresponding to the post.
 func (a *App) FillInPostProps(c request.CTX, post *model.Post, channel *model.Channel) *model.AppError {
 	channelMentions := post.ChannelMentions()
+	ChannelMentionsAcrossTeams := post.ChannelMentionsAcrossTeams()
```
Comment (agnivade) [inline server/channels/app/post.go:464]: Let's also add a test similar to `TestPostChannelMentions` to test this functionality and verify that the post prop gets set properly.

File: server/channels/app/post.go [hunk 4]
```diff
@@ -492,6 +493,28 @@ func (a *App) FillInPostProps(c request.CTX, post *model.Post, channel *model.Ch
 		}
 	}
 
+	if len(ChannelMentionsAcrossTeams) > 0 {
+		for mentionedTeam, mentionedTeamChannels := range ChannelMentionsAcrossTeams {
+			team, err := a.Srv().Store().Team().GetByName(mentionedTeam)
+			if err != nil {
+				mlog.Warn("Failed to get team of the channel mention", mlog.String("mentionedTeam", channel.TeamId))
```
Comment (agnivade) [inline server/channels/app/post.go:500]: Use `c.Logger().Warn(` instead of `mlog.Warn`. Let's do the change in L485 as well.

**Review by agnivade — COMMENTED** [review]
Thanks Dhama. Left some comments.

File: webapp/channels/src/utils/text_formatting.tsx [hunk 5]
```diff
@@ -512,15 +512,16 @@ export function autolinkChannelMentions(
...
-                value: `<a class="mention-link" href="${href}" data-channel-mention-team="${teamName}" data-channel-mention="${channelName}">~${displayName}</a>`,
+                value: `<a class="mention-link" href="${href}" data-channel-mention-team="${teamName}" data-channel-mention="${channelName}">~${displayName}${teamTxt}</a>`,
                 originalText: mention,
             });
         } else if (team) {
             // eslint-disable-next-line @typescript-eslint/no-explicit-any
             href = ((window as any).basename || '') + '/' + team.name + '/channels/' + channelName;
             tokens.set(alias, {
-                value: `<a class="mention-link" href="${href}" data-channel-mention="${channelName}">~${displayName}</a>`,
+                value: `<a class="mention-link" href="${href}" data-channel-mention="${channelName}" data-channel-mention-team="${team.name}">~${displayName}</a>`,
```
Comment (hmhealey) [inline webapp/channels/src/utils/text_formatting.tsx:524]: Would we also want the team name to be displayed here? I can't remember the difference between these two cases

**Review by hmhealey — CHANGES_REQUESTED** [review]
Thanks, Ashish!

One thing I'm not sure about with this is the `~channel(team)` format for these in the text. I think it's the regular brackets that have me a bit worried because people might use those when writing text normally. I saw `~channel:team` suggested above which does look a bit cleaner, but that might run into the same issue as https://mattermost.atlassian.net/browse/MM-34627 where the Markdown parser tries to treat it as a URL instead of plain text because of the `:`. Ideally, we'd encode them as something like `~channelId` since then we wouldn't need the team name at all, but that drastic of a change leads to more compatibility concerns and poor UX if we don't have a richer post editor by that point.

Thoughts on potentially changing that? I haven't seen any discussion on that previously, so I'm curious about your thinking here too.

Also, speaking of compatibility concerns, we'll also need to support mobile with this as well. Would you be interested in working on that too, or should we split that off into a separate ticket?

**Comment by unode** [comment]
Was this accidentally closed?

**Comment by AshishDhama** [comment]
I discussed with @hmhealey, that since the current solution isn't taking care of all the cases I am closing this pr.

**Comment by unode** [comment]
@AshishDhama would it be possible to have a summary of the cases?
Perhaps someone can then craft another PR that addresses all points.

**Comment by hmhealey** [comment]
It's a bit hard to explain since there's a lot of history here, but the short version of this is that cross-team channel linking is a lot larger effort than initially expected because those short links were written before MM users could be on multiple teams. There's issues around both permissions, caching, and Markdown, and the Markdown changes in particular introduce compatibility issues with mobile which would also need to be solved, so we have to think more about when we'll be able to do this

**Comment by unode** [comment]
Thanks @hmhealey. Personally I still see this feature as quite important to users, particularly new ones to the MM ecosystem. For instance, in a direct message, the meaning of a channel link `~Town Square` changes depending on which team you have active when reading the message or clicking the link. This regularly causes confusion and messages ending up in the wrong channels.

I would be happy to contribute if there's any direction that seems promising or if there is space to propose complete or partial solutions. Possibly even a series of changes to be slowly introduced over several versions.

Specifically, I see two sub-tasks (in order of relevance):
1. Syntax for cross-team channel reference (possibly even cross-instance) that is compatible with current markdown parsers and unambiguously refers to a channel.
2. Auto-complete behavior (by channel name before team? by team before channel? public channels only? private channels that the user is a member of?)

**Pull request closed without merging**